### PR TITLE
Feat  icp proxy

### PR DESCRIPTION
**Pull request by alenmestrov**
# ICP Proxy Contract Implementation

## Summary
This PR implements the Internet Computer (ICP) version of the proxy contract, enabling multi-signature management of cross-canister calls and contract administration.

## Key Features
- Multi-signature proposal system
- Support for various proposal actions:
  - External function calls to any canister
  - ICP token transfers
  - Contract configuration updates (num approvals, proposal limits)
  - Context value management
- Proposal lifecycle management with approval tracking
- Secure signature verification

### Core Implementation
- Added ICP-specific contract implementation using Candid types
- Implemented core types matching NEAR's structure:
  - Base types: `ICSignerId`, `ICProposalId`
  - Proposal types: `ICProposal`, `ICProposalAction`
  - Request types: `ICRequest`, `ICRequestKind`
  - Contract state management

### Integration Tests
Added comprehensive test suite covering:
- Proposal creation and execution
- Multi-signature approval flow
- External function calls
- Token transfers
- Configuration updates
- Edge cases and error handling

### Prerequisites
- DFX installed
- Rust with wasm32-unknown-unknown target
- candid-extractor: `cargo add candid-extractor`
### Prerequisites
1. Install Rust and add wasm target:
```bash
rustup target add wasm32-unknown-unknown
```

2. Install DFX (Internet Computer SDK):
```bash
sh -ci "$(curl -fsSL https://internetcomputer.org/install.sh)"
```

3. Add candid-extractor for .did file generation:
```bash
cargo add candid-extractor
```

4. Download and install pocket-ic:
   - Download from: https://github.com/dfinity/pocketic/releases
   - Place the binary in the `proxy-contract` directory
   - Make it executable: `chmod +x pocket-ic`
   
### Mock Contracts
The project includes mock contracts for testing purposes:

1. **Mock Ledger Contract** (\`mock/ledger\`):
   - Simulates the ICP ledger functionality
   - Handles token transfers
   - Maintains balance state
   - Used for testing transfer proposal actions

2. **Mock External Contract** (\`mock/external\`):
   - Generic test contract for external calls
   - Records received calls and arguments
   - Used for testing external function call proposal actions

### Build Process
1. First, build the mock contracts using the provided script:
```bash
./build_mock_contracts.sh
```
This script:
- Builds both mock contracts to wasm
- Generates their .did files
- Must be run before running tests

2. Then build the main contract:
```bash
./build.sh
```
This script:
- Check if the wasm32 target is installed
- Builds the Proxy contract
- Generate .did file
- Stop local environment if running
- Start local environment in the background
- Deploy the contract on the local environment 

### Running Tests

To run the integration tests:
```bash
cargo test --test integration -- --nocapture
```

### Test Structure
The integration tests cover:
- Proposal creation and validation
- Multi-signature approval flow
- External function calls to other canisters
- Token transfers
- Contract configuration updates
- Error cases and edge conditions

## Notes
- External function calls accept hex-encoded arguments to support any canister interface
- Proposals require multiple approvals based on contract configuration
- Contract maintains proposal state and approval tracking
- Integration with ICP ledger for token transfers

### CODE REVIEW

File: contracts/icp/proxy-contract/src/types.rs [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    pub context_id: ICContextId,
+    pub context_config_id: String,
+    pub num_approvals: u32,
+    pub proposals: HashMap<ICProposalId, ICProposal>,
+    pub approvals: HashMap<ICProposalId, HashSet<ICSignerId>>,
+    pub num_proposals_pk: HashMap<ICSignerId, u32>,
+    pub active_proposals_limit: u32,
+    pub context_storage: HashMap<Vec<u8>, Vec<u8>>,
+    pub ledger_id: LedgerId,
+}
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/types.rs]: can you move contract types to lib.rs? they have no value used elsewhere, but the rest of the types do.

File: contracts/icp/proxy-contract/src/types.rs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+    fn idl_serialize<S>(&self, serializer: S) -> Result<(), S::Error>
+    where
+        S: candid::types::Serializer,
+    {
+        serializer.serialize_null(())
+    }
+}
+
+#[derive(CandidType, Serialize, Deserialize, Default)]
+pub struct ICProxyContract {
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/types.rs]: this one derives `CandidType`, but `ContextConfigs` didn't, what requires it?

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+            let num_proposals = PROXY_CONTRACT.with(|contract| {
+                let contract = contract.borrow();
+                contract
+                    .num_proposals_pk
+                    .get(&proposal.author_id)
+                    .copied()
+                    .unwrap_or(0)
+            });
+
+            internal_create_proposal(proposal.clone(), num_proposals)
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: `internal_create_proposal` already reads `PROXY_CONTRACT.with`, why not fetch it there?

```suggestion
            internal_create_proposal(proposal.clone()
```

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    if current_time.saturating_sub(request.timestamp_ms) > 1000 * 5 {
+        return Err("request expired".to_string());
+    }
+
+    // Check membership
+    if !check_member(&request.signer_id).await? {
+        return Err("signer is not a member".to_string());
+    }
+
+    match &request.kind {
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]:
```suggestion
    match request.kind {
```

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+            });
+
+            internal_create_proposal(proposal.clone(), num_proposals)
+        }
+        ICRequestKind::Approve { approval } => {
+            internal_approve_proposal(
+                approval.signer_id.clone(),
+                approval.proposal_id.clone(),
+                approval.added_timestamp,
+            )
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: so you don't need to clone

```suggestion
            internal_create_proposal(proposal, num_proposals)
        }
        ICRequestKind::Approve { approval } => {
            internal_approve_proposal(
                approval.signer_id,
                approval.proposal_id,
                approval.added_timestamp,
            )
```

File: contracts/icp/proxy-contract/src/types.rs [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+    }
+}
+
+impl Default for ICContextId {
+    fn default() -> Self {
+        Self(Identity::default())
+    }
+}
+
+#[derive(CandidType, Serialize, Deserialize, Clone, Debug, Hash, Eq, PartialEq)]
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/types.rs:122]:
```suggestion
#[derive(CandidType, Serialize, Deserialize, Copy, Clone, Debug, Hash, Eq, PartialEq)]
```

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+    // First phase: Update approvals and check if we need to execute
+    let should_execute = PROXY_CONTRACT.with(|contract| {
+        let mut contract = contract.borrow_mut();
+
+        // Check if proposal exists
+        if !contract.proposals.contains_key(&proposal_id) {
+            return Err("proposal does not exist".to_string());
+        }
+
+        let approvals = contract.approvals.entry(proposal_id.clone()).or_default();
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: so you can..

```suggestion
        let approvals = contract.approvals.entry(proposal_id).or_default();
```

File: contracts/icp/proxy-contract/src/types.rs [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+#[derive(CandidType, Serialize, Deserialize, Default)]
+pub struct ICProxyContract {
+    pub context_id: ICContextId,
+    pub context_config_id: String,
+    pub num_approvals: u32,
+    pub proposals: HashMap<ICProposalId, ICProposal>,
+    pub approvals: HashMap<ICProposalId, HashSet<ICSignerId>>,
+    pub num_proposals_pk: HashMap<ICSignerId, u32>,
+    pub active_proposals_limit: u32,
+    pub context_storage: HashMap<Vec<u8>, Vec<u8>>,
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/types.rs]: let's use `BTree{Set,Map}`, unless you're convinced hashing keys in this case is more optimal

IMO, 32-byte key comparison is pretty cheap for btree operations, compared to applying a hash function over the keys

well, except for `context_storage`, that can stay a `HashMap`

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+
+        if approvals.contains(&signer_id) {
+            return Err("proposal already approved".to_string());
+        }
+
+        approvals.insert(signer_id);
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: alternatively;

```suggestion
        if !approvals.insert(&signer_id) {
            return Err("proposal already approved".to_string());
        }
```

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+            ICProposalAction::ExternalFunctionCall {
+                receiver_id,
+                method_name,
+                args,
+                deposit: _,
+            } => {
+                let args_bytes =
+                    hex::decode(args).map_err(|e| format!("Invalid args hex encoding: {}", e))?;
+
+                let _: () = ic_cdk::call(receiver_id, method_name.as_str(), (args_bytes,))
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: ah, I thought `ic_call` would take the hex-encoded string, if it takes a `Vec<u8>` instead, then why are we hex-encoding?

Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: can we guarantee the returned type WILL be `null`?

otherwise, should we decode to `IDLValue`?

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+                // First encode to bytes
+                let args_bytes =
+                    candid::encode_one(transfer_args).expect("Failed to encode transfer args");
+
+                // Then wrap in newtype struct like the working version
+                #[derive(CandidType)]
+                struct Args(Vec<u8>);
+
+                let _: () =
+                    ic_cdk::call(Principal::from(ledger_id), "transfer", (Args(args_bytes),))
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: do we need to wrap it in args? you didn't for external-function call

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+        }
+    }
+    Ok(())
+}
+
+fn remove_proposal(proposal_id: ICProposalId) {
+    PROXY_CONTRACT.with(|contract| {
+        let mut contract = contract.borrow_mut();
+        contract.approvals.remove(&proposal_id);
+        if let Some(proposal) = contract.proposals.remove(&proposal_id) {
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]:
```suggestion
fn remove_proposal(proposal_id: &ICProposalId) {
    PROXY_CONTRACT.with(|contract| {
        let mut contract = contract.borrow_mut();
        contract.approvals.remove(proposal_id);
        if let Some(proposal) = contract.proposals.remove(proposal_id) {
```

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+            ICProposalAction::SetContextValue { key, value } => {
+                PROXY_CONTRACT.with(|contract| {
+                    let mut contract = contract.borrow_mut();
+                    contract.context_storage.insert(key.clone(), value.clone());
+                });
+            }
+        }
+    }
+
+    remove_proposal(proposal_id.clone());
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: so this can be

```suggestion
    remove_proposal(proposal_id);
```

File: contracts/icp/proxy-contract/src/mutate.rs (hunk 11 above)
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: here too

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+            method_name,
+            args,
+            deposit: _,
+        } => {
+            if method_name.is_empty() {
+                return Err("method name cannot be empty".to_string());
+            }
+            if args.is_empty() {
+                return Err("args cannot be empty".to_string());
+            }
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: ICP doesn't permit this?

how about functions that don't take any arguments, do they still need a non-empty `args` definition?

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+            .insert(proposal_id.clone(), proposal.clone());
+
+        // Initialize approvals set with author's approval
+        let mut approvals = HashSet::new();
+        approvals.insert(proposal.author_id.clone());
+        contract.approvals.insert(proposal_id.clone(), approvals);
+
+        // Update proposal count
+        let author_id = proposal.author_id;
+        *contract.num_proposals_pk.entry(author_id).or_insert(0) += 1;
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: it's cheaper to copy author ID than clone proposal ID

```suggestion
        let proposal_id = proposal.id;
        let author_id = proposal.author_id;

        contract
            .proposals
            .insert(proposal_id, proposal);

        // Initialize approvals set with author's approval
        let mut approvals = HashSet::from([author_id]);
        contract.approvals.insert(proposal_id, approvals);

        *contract.num_proposals_pk.entry(author_id).or_default() += 1;
```

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+        ICProposalAction::Transfer {
+            receiver_id: _,
+            amount,
+        } => {
+            if *amount == 0 {
+                return Err("transfer amount cannot be zero".to_string());
+            }
+
+            if *amount > 1_000_000_000 {
+                return Err("transfer amount limit exceeded".to_string());
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: what is this limit? should it be static?

File: contracts/icp/proxy-contract/src/mutate.rs [hunk 17]
```diff
@@ -0,0 +1,304 @@
...
+                return Err("active proposals limit cannot be zero".to_string());
+            }
+        }
+        ICProposalAction::SetContextValue { key, value } => {
+            if key.is_empty() {
+                return Err("key cannot be empty".to_string());
+            }
+            if value.is_empty() {
+                return Err("value cannot be empty".to_string());
+            }
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/mutate.rs]: any reason why not?

File: contracts/icp/proxy-contract/src/types.rs
Comment (miraclx) [inline contracts/icp/proxy-contract/src/types.rs:1]: we shouldn't need to duplicate most of the code here, why not import from context-config types?

File: contracts/icp/proxy-contract/src/query.rs [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+        })
+    })
+}
+
+#[ic_cdk::query]
+pub fn get_proposal_approvers(proposal_id: ICProposalId) -> Option<Vec<ICSignerId>> {
+    PROXY_CONTRACT.with(|contract| {
+        let contract = contract.borrow();
+        if let Some(approvals) = contract.approvals.get(&proposal_id) {
+            Some(approvals.iter().flat_map(|a| a.rt()).collect())
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/query.rs]: do we need `ReprTransmute` here?

File: contracts/icp/proxy-contract/src/mutate.rs (hunk 17 above)
Comment (alenmestrov) [reply contracts/icp/proxy-contract/src/mutate.rs]: I put this check here since it if the key or value that we want to add are empty it should fail right?

File: contracts/icp/proxy-contract/src/types.rs
Comment (alenmestrov) [reply contracts/icp/proxy-contract/src/types.rs:1]: I wanted to keep it seperate for each contract for now. I can use types from Context contract if you think its ok.

File: contracts/icp/proxy-contract/src/mutate.rs (hunk 14 above)
Comment (alenmestrov) [reply contracts/icp/proxy-contract/src/mutate.rs]: the ICP requirement is to encode the arguments for the call. They can be empty but currently, I can't think of a call that wont be needing arguments.

File: contracts/icp/proxy-contract/src/types.rs (hunk 2 above)
Comment (alenmestrov) [reply contracts/icp/proxy-contract/src/types.rs]: CandidType is required when a type needs to be serialized/deserialized for inter-canister communication or when it's part of a canister's public interface (input/output of public methods).

Since we are exposing ICProxyContract directly to canister's public interface we need it here.

File: contracts/icp/proxy-contract/src/mutate.rs (hunk 10 above)
Comment (alenmestrov) [reply contracts/icp/proxy-contract/src/mutate.rs]: resolved

File: contracts/icp/proxy-contract/src/mutate.rs (hunk 11 above)
Comment (alenmestrov) [reply contracts/icp/proxy-contract/src/mutate.rs]: resolved

File: contracts/icp/proxy-contract/src/types.rs (hunk 2 above)
Comment (miraclx) [reply contracts/icp/proxy-contract/src/types.rs]: but didn't we expose `ContextConfigs`? if not, then why do we expose this one? they're both canister state, right?

File: contracts/icp/proxy-contract/src/mutate.rs (hunk 14 above)
Comment (miraclx) [reply contracts/icp/proxy-contract/src/mutate.rs]: since this is an external function call, we cannot preemptively know if the function will have any arguments, can you test introducing a function with no arguments and see if calling it works with no args.

File: contracts/icp/proxy-contract/src/mutate.rs (hunk 17 above)
Comment (miraclx) [reply contracts/icp/proxy-contract/src/mutate.rs]: we don't have that limitation on NEAR, or starknet? I don't think. And I can't think of a reason to have it.

File: contracts/icp/proxy-contract/src/types.rs
Comment (miraclx) [reply contracts/icp/proxy-contract/src/types.rs:1]: yeah, lets.. proxy contract and context config are inextricably bound anyway, makes sense.

File: contracts/icp/proxy-contract/src/mutate.rs (hunk 17 above)
Comment (alenmestrov) [reply contracts/icp/proxy-contract/src/mutate.rs]: ok fair enough, will remove it

File: contracts/icp/proxy-contract/src/types.rs
Comment (alenmestrov) [reply contracts/icp/proxy-contract/src/types.rs:1]: will do that in final PR

File: contracts/icp/proxy-contract/src/mutate.rs (hunk 17 above)
Comment (alenmestrov) [reply contracts/icp/proxy-contract/src/mutate.rs]: removed

File: contracts/icp/proxy-contract/src/types.rs (hunk 2 above)
Comment (alenmestrov) [reply contracts/icp/proxy-contract/src/types.rs]: True, I just tested it without CandidType and it works.

File: contracts/icp/proxy-contract/src/lib.rs [hunk 19]
```diff
@@ -14,7 +14,7 @@ pub mod mutate;
 pub mod query;
 pub mod types;
 
-#[derive(CandidType, Serialize, Deserialize, Default)]
+#[derive(Serialize, Deserialize, Default)]
```
Comment (miraclx) [inline contracts/icp/proxy-contract/src/lib.rs]: everything but `Default` can be removed, yeah?

```suggestion
#[derive(Default)]
```

**Review by miraclx — APPROVED** [review]

File: scripts/test.sh [hunk 20]
```diff
@@ -13,6 +13,7 @@ cd "$(dirname $0)"
 ../contracts/context-config/build.sh
 ../contracts/proxy-lib/build-test-deps.sh
 ../contracts/icp/context-config/build.sh
-
+../contracts/icp/proxy-contract/build_mock_contracts.sh
+../contracts/icp/proxy-contract/build.sh
```
Comment (miraclx) [inline scripts/test.sh]: `build_mock_contracts` also calls `proxy-contract/build.sh`, maybe we skip the bottom one?

**Review by miraclx — APPROVED** [review]

File: scripts/test.sh (hunk 20 above)
Comment (alenmestrov) [reply scripts/test.sh]: true, will fix it in final PR
